New song select footer design implementation

mk56-spn opened the pull request:
supersedes #20148 and is part of  https://github.com/ppy/osu/issues/18890.
- Addresses https://github.com/ppy/osu/discussions/20135
- Resolves https://github.com/ppy/osu/issues/5923

Cleaner commit history and isolated test only implementation to allow for merging without concern for other components it needs to work with.

line count is majorly bloated due to it being made as a drag and drop replacement, which is also the reason I put it all in one folder

mk56-spn committed:
Create Basic V2 footer and test

mk56-spn committed:
Add basic button design and footer button addition flow

mk56-spn committed:
Add button "accent" colour, bottom bar, icon, text

mk56-spn committed:
Add functionality of Random button

mk56-spn committed:
Add hover lightening

mk56-spn committed:
Add disabled button dimming.

mk56-spn committed:
Add button colouring whilst corresponding overlay is present

mk56-spn committed:
Use OsuColour instead of hex for button accent colour

Joehuu commented on osu.Game/Screens/Select/FooterV2/FooterButtonModsV2.cs:
```diff
@@ -0,0 +1,20 @@
+        private void load(OsuColour colour)
+        {
+            Text = "Mods";
+            Icon = FontAwesome.Solid.ArrowsAlt;
```
If we are still following the figma designs, this should be:
```suggestion
            Icon = FontAwesome.Solid.ExchangeAlt;
```

mk56-spn replied:
my bad, checked the wrong commit on the old branch

mk56-spn commented on osu.Game/Screens/Select/FooterV2/FooterV2.cs:
```diff
@@ -0,0 +1,91 @@
+            };
+        }
+
+        protected override bool OnMouseDown(MouseDownEvent e) => true;
+
+        protected override bool OnClick(ClickEvent e) => true;
+
+        protected override bool OnHover(HoverEvent e) => true;
```
good catch

mk56-spn commented on osu.Game/Screens/Select/FooterV2/FooterV2.cs:
```diff
@@ -0,0 +1,91 @@
+{
+    public partial class FooterV2 : Container
+    {
+        //Should be 60, setting to 50 for now for the sake of matching the current BackButton height.
```
yeah gonna wait until new back button is added, whether that be after or before this PR is merged since its a quick change anyways

mk56-spn commented on osu.Game/Screens/Select/FooterV2/FooterButtonV2.cs:
```diff
@@ -0,0 +1,212 @@
+{
+    public partial class FooterButtonV2 : OsuClickableContainer, IKeyBindingHandler<GlobalAction>
+    {
+        private const int button_height = 120;
```
i used 120 because i was using size changing transforms initially, 120 seems rather sane though since its just adding 50% wherever needed

mk56-spn commented on osu.Game/Screens/Select/FooterV2/FooterButtonV2.cs:
```diff
@@ -0,0 +1,212 @@
+        private const int corner_radius = 10;
+        private const int transition_length = 500;
+
+        public const float SHEAR_WIDTH = 16;
```
yep. wasnt aware of being able to check the exact shear with a copy

mk56-spn committed:
Address issues from Joehuu review

mk56-spn committed:
Merge branch 'master' into footer_V2_implementation

mk56-spn committed:
fix test failures, improve button logic

mk56-spn committed:
Make offsetting of bottom bar more readable

mk56-spn commented on osu.Game/Screens/Select/FooterV2/FooterButtonV2.cs:
```diff
@@ -0,0 +1,203 @@
+                        new Container
+                        {
+                            //Offset the bar to centre it with consideration for the shearing
+                            Position = new Vector2(-0.15f * 35, -10),
```
updated to be cleaner, i think this way is better

Joehuu replied:
I still think using moving the container out and using `Shear = -SHEAR` is better because it's more self-explanatory and doesn't need a comment.

mk56-spn replied:
Will do

mk56-spn committed:
adjust colourbar to position itself in a more simple fashion

peppy commented on osu.Game/Screens/Select/FooterV2/FooterButtonFreeModsV2.cs:
```diff
@@ -0,0 +1,21 @@
+
+namespace osu.Game.Screens.Select.FooterV2
+{
+    public partial class FooterButtonFreeModsV2 : FooterButtonV2
```
This isn't displayed anywhere?

mk56-spn replied:
im aware it isn't, thought it wasn't worth adding to the test scene since it has no special functionality and wasnt present on the test this test is replacing, and its usage case is clear, but ill add it now

peppy replied:
We don't add classes to the project unless they have a usage.

peppy commented on osu.Game/Screens/Select/FooterV2/FooterButtonV2.cs:
```diff
@@ -0,0 +1,195 @@
+
+        protected override void OnHoverLost(HoverLostEvent e) => updateDisplay();
+
+        protected override bool OnMouseDown(MouseDownEvent e) => !Enabled.Value || base.OnMouseDown(e);
+        protected override bool OnClick(ClickEvent e) => !Enabled.Value || base.OnClick(e);
```
I'm very confused by these conditionals. I do not think they should exist.

mk56-spn replied:
what about them to be exact? i took this design pattern from the old footer buttons that were recently updated. although i did inline it prompted by the jetbrains tips, if thats the issue i can revert it, just making sure

peppy replied:
I'm not sure what the purpose is. Can you explain the purpose? If you can't, then it's probably not a good sign.

One rule to follow is to not blindly copy code, even if you are working with existing code. I believe these conditionals are going to make mouse clicks travel *through* the button to elements behind them when the buttons are not enabled, which is probably incorrect. I'd expect them to always be blocking mouse down / clicks?

mk56-spn committed:
add free mods button to test

mk56-spn committed:
Remove ```onclick``` and ```mousedown``` override

mk56-spn committed:
Remove free mods button

peppy commented on osu.Game/Screens/Select/FooterV2/FooterButtonV2.cs:
```diff
@@ -0,0 +1,194 @@
+                },
+                new Container
+                {
+                    // The X offset has to multiplied as such to account for the fact that we only want to offset by the distance from the CenterLeft point of the container
```
Which X offset? There's none specified.

mk56-spn replied:
Oversight. Referenced some jank positioning I had before . Will remove in a bit when I can get to my pc

mk56-spn committed:
Remove leftover comment regarding X axis offsets

peppy commented on osu.Game/Screens/Select/FooterV2/FooterV2.cs:
```diff
@@ -0,0 +1,86 @@
+                new Box
+                {
+                    RelativeSizeAxes = Axes.Both,
+                    Colour = colour.B5
```
Not sure about this one. But it matches the design so let's go with it for now. I've left a comment in figma questioning its validity.

Joehuu replied:
This should be `colourProvider.Background5`.

peppy committed:
Adjust footer button colour handling to read better and take into account mouse down

peppy committed:
Use `FinishTransforms` rather than manual duplication of background colour logic

peppy committed:
Adjust comments and formatting of comments

peppy committed:
Merge branch 'master' into footer_V2_implementation

peppy reviewed (dismissed):
Seems okay for a first pass.

Joehuu commented on osu.Game/Screens/Select/FooterV2/FooterV2.cs:
```diff
@@ -0,0 +1,86 @@
+                    RelativeSizeAxes = Axes.Both,
+                    Colour = colour.B5
+                },
+                new FillFlowContainer
```
This fillflow is not needed.

Joehuu commented on osu.Game/Screens/Select/FooterV2/FooterButtonV2.cs:
```diff
@@ -0,0 +1,210 @@
+            {
+                Type = EdgeEffectType.Shadow,
+                Radius = 5,
+                Roundness = 10,
```
Not sure if roundness is needed, see xmldoc:
> Adds to the CornerRadius of the corresponding CompositeDrawable.
If it is:
```suggestion
                Roundness = corner_radius,
```

Joehuu commented on osu.Game/Screens/Select/FooterV2/FooterButtonV2.cs:
```diff
@@ -0,0 +1,210 @@
+            EdgeEffect = new EdgeEffectParameters
+            {
+                Type = EdgeEffectType.Shadow,
+                Radius = 5,
```
Figma says 4 blur radius. If it doesn't match, it should have a comment like all the other ones.

Joehuu replied:
This wasn't applied or commented:

Leaving this suggestion here if it matches or is unnoticable:
```suggestion
                Radius = 4,
```

mk56-spn replied:
Yeah sorry thought I'd applied it for some reason

Joehuu commented on osu.Game/Screens/Select/FooterV2/FooterButtonV2.cs:
```diff
@@ -0,0 +1,210 @@
+                Type = EdgeEffectType.Shadow,
+                Radius = 5,
+                Roundness = 10,
+                Colour = Colour4.Black.Opacity(0.25f)
```
Figma says 50%. ^

There is also an offset:
```
                Colour = Colour4.Black.Opacity(0.5f),
                Offset = new Vector2(0, 2),
```

Joehuu commented on osu.Game/Screens/Select/FooterV2/FooterButtonV2.cs:
```diff
@@ -0,0 +1,210 @@
+                    Shear = -SHEAR,
+                    Anchor = Anchor.BottomCentre,
+                    Origin = Anchor.Centre,
+                    Y = -10,
```
```suggestion
                    Y = -corner_radius,
```

mk56-spn committed:
Change FooterV2.cs colour to use `ColourProvider` instead of `OsuColour`.

Remove unnecessary `FillFlowContainer`

mk56-spn committed:
Fix corner_radius missing in some parts of `FooterButtonV2.cs`.

Adjust shadow radius value to 5 to match figma.

mk56-spn committed:
Add note explaining shadow opacity

pass ColourProvider in from test, instead of hard coding it in `FooterButtonV2.cs`

mk56-spn committed:
Add offset to `FooterButtonV2.cs` shadow

mk56-spn committed:
Update osu.Game/Screens/Select/FooterV2/FooterButtonV2.cs

Co-authored-by: Joseph Madamba <madamba.joehu@outlook.com>

Joehuu reviewed (dismissed):
LGTM

mk56-spn commented:
thanks for all the nitpicks btw joehuu, i need to be more careful

peppy committed:
Make `Action`s nullable

peppy commented on osu.Game/Screens/Select/FooterV2/FooterButtonRandomV2.cs:
```diff
@@ -0,0 +1,161 @@
+    public partial class FooterButtonRandomV2 : FooterButtonV2
+    {
+        public Action NextRandom { get; set; } = null!;
+        public Action PreviousRandom { get; set; } = null!;
```
Generally we always make `Action`s `Action?` and support nullability, as a pattern.

peppy committed:
Merge branch 'master' into footer_V2_implementation
